### ISSUE TITLE
Error when trying to create mixed system

### ISSUE THREAD

**Issue by victorprincipe** [openmm/openmm-torch#27]
Hi,

I've been trying to use ANI-2x to model the ligand during the simulation of a protein-ligand complex. I've been using the code below to set up the system.

```
prmtop = AmberPrmtopFile(prmtop_file)
inpcrd = AmberInpcrdFile(crd_file)

potential = MLPotential('ani2x')
ml_atoms  = [atom.index for atom in prmtop.topology.atoms() if atom.residue.name == "MOL"]
mm_system = prmtop.createSystem(nonbondedMethod=PME,
nonbondedCutoff=1.0*unit.nanometers, constraints=HBonds, rigidWater=True,
ewaldErrorTolerance=0.0005)
ml_system = potential.createMixedSystem(prmtop.topology, mm_system, ml_atoms)
```
The last line in the code raises the following error:

```
Traceback (most recent call last):
  File "/Users/victorprincipe/Local/project2021/local_tests/openmm_ani_test/md_ani.py", line 26, in <module>
    ml_system = potential.createMixedSystem(prmtop.topology, mm_system, ml_atoms)
  File "/Users/victorprincipe/miniconda3/envs/openmmani/lib/python3.9/site-packages/openmmml-1.0-py3.9.egg/openmmml/mlpotential.py", line 278, in createMixedSystem
    self._impl.addForces(topology, newSystem, atomList, forceGroup, **args)
  File "/Users/victorprincipe/miniconda3/envs/openmmani/lib/python3.9/site-packages/openmmml-1.0-py3.9.egg/openmmml/models/anipotential.py", line 85, in addForces
    species = model.species_to_tensor(elements).unsqueeze(0)
  File "/Users/victorprincipe/miniconda3/envs/openmmani/lib/python3.9/site-packages/torchani/models.py", line 165, in species_to_tensor
    return self._species_to_tensor(*args, **kwargs) \
  File "/Users/victorprincipe/miniconda3/envs/openmmani/lib/python3.9/site-packages/torchani/utils.py", line 237, in __call__
    rev = [self.rev_species[s] for s in species]
  File "/Users/victorprincipe/miniconda3/envs/openmmani/lib/python3.9/site-packages/torchani/utils.py", line 237, in <listcomp>
    rev = [self.rev_species[s] for s in species]
KeyError: 'l'
```
 
I'm not sure why this is occurring, so any help would be greatly appreciated. I've also attached the prmtop and inpcrd files below. Thanks in advance!

[system_files.zip](https://github.com/openmm/openmm-torch/files/6156373/system_files.zip)

**Comment by peastman** [comment]
It looks like TorchANI has changed the input to `species_to_tensor()`.  It used to be a string listing the element symbols.  That was fine for ANI1, which only supported elements with one-character symbols.  ANI2 adds Cl, and apparently it doesn't handle the two-character symbol correctly.  Instead they allow you to pass a list of symbols rather than a single string of concatenated symbols.

Let me fix that.  It's a trivial change.

**Comment by peastman** [comment]
The fix is in https://github.com/openmm/openmm-ml/pull/7.  Could you try it out and verify it fixes the problem for you?

**Comment by victorprincipe** [comment]
I just tried it and the change solved the issue! Thank you for looking into it for me. Unfortunately I get another error message now during the minimisation step (shown below). Do you know what could be causing this issue?

```
Traceback (most recent call last):
  File "/Users/victorprincipe/Local/project2021/local_tests/openmm_ani_test/md_ani.py", line 38, in <module>
    simulation.minimizeEnergy()
  File "/Users/victorprincipe/miniconda3/envs/openmmani/lib/python3.9/site-packages/simtk/openmm/app/simulation.py", line 128, in minimizeEnergy
    mm.LocalEnergyMinimizer.minimize(self.context, tolerance, maxIterations)
  File "/Users/victorprincipe/miniconda3/envs/openmmani/lib/python3.9/site-packages/simtk/openmm/openmm.py", line 12810, in minimize
    return _openmm.LocalEnergyMinimizer_minimize(context, tolerance, maxIterations)
simtk.openmm.OpenMMException: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
  File "<string>", line 3, in forward

      def addmm(self: Tensor, mat1: Tensor, mat2: Tensor, beta: number = 1.0, alpha: number = 1.0):
          return self + mat1.mm(mat2)
                        ~~~~~~~ <--- HERE

      def batch_norm(input : Tensor, running_mean : Optional[Tensor], running_var : Optional[Tensor], training : bool, momentum : float, eps : float) -> Tensor:
RuntimeError: expected scalar type Double but found Float
```

**Comment by peastman** [comment]
Oops, I though I had already fixed that, but apparently I didn't actually check it in.  The fix is in https://github.com/openmm/openmm-ml/pull/8.

**Comment by victorprincipe** [comment]
Thank you, that fix solved it. Now getting errors when setting velocities to a temperature due to the autogrid engine being called while holding the GIL, which seems to be an error from pytorch so I'll raise it on there.
Thanks again for all the help!

**Comment by peastman** [comment]
Fixing that required a change to OpenMM.  It's in https://github.com/openmm/openmm/pull/3061.  The fix will be in 7.5.1, which should be released shortly.

**Comment by victorprincipe** [comment]
Ah ok I'll wait on the fix from OpenMM then, thanks for letting me know

**Comment by peastman** [comment]
It's also in the latest nightly build, if you want to start experimenting with it now.

```
conda install -c conda-forge -c omnia-dev openmm
```

Just remember that nightly builds aren't well tested and should never be used for production simulations.

**Comment by victorprincipe** [comment]
I've installed the nightly build into a new environment to test it out, but I'm running into some other errors on both my home computer (running macOS) and a remote work station I use (running Linux). For the Linux system I am getting a segmentation fault with no other information:
`Segmentation fault (core dumped)`
For the macOS system (using the CPU as the platform), I am getting an error regarding an unknown device:
`Traceback (most recent call last):
  File "/Users/victorprincipe/Local/project2021/local_tests/openmm_ani_test/md_ani.py", line 33, in <module>
    simulation = Simulation(prmtop.topology, ml_system, integrator, platform)
  File "/Users/victorprincipe/miniconda3/envs/openmmanidev/lib/python3.9/site-packages/simtk/openmm/app/simulation.py", line 103, in __init__
    self.context = mm.Context(self.system, self.integrator, platform)
  File "/Users/victorprincipe/miniconda3/envs/openmmanidev/lib/python3.9/site-packages/simtk/openmm/openmm.py", line 2678, in __init__
    _openmm.Context_swiginit(self, _openmm.new_Context(*args))
simtk.openmm.OpenMMException: Unknown device: -11. If you have recently updated the caffe2.proto file to add a new device type, did you forget to update the DeviceTypeName() function to reflect such recent changes?
Exception raised from DeviceTypeName at /tmp/pip-req-build-4_wdsgbr/c10/core/DeviceType.cpp:42 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) + 112 (0x1141ab030 in libc10.dylib)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 106 (0x1141a990a in libc10.dylib)
frame #2: c10::DeviceTypeName(c10::DeviceType, bool) + 598 (0x114192b06 in libc10.dylib)
frame #3: torch::jit::Unpickler::readInstruction() + 12485 (0x11ab57575 in libtorch_cpu.dylib)
frame #4: torch::jit::Unpickler::run() + 312 (0x11ab54338 in libtorch_cpu.dylib)
frame #5: torch::jit::Unpickler::parse_ivalue() + 43 (0x11ab5dd3b in libtorch_cpu.dylib)
frame #6: torch::jit::readArchiveAndTensors(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, c10::optional<std::__1::function<c10::StrongTypePtr (c10::QualifiedName const&)> >, c10::optional<std::__1::function<c10::intrusive_ptr<c10::ivalue::Object, c10::detail::intrusive_target_default_null_type<c10::ivalue::Object> > (c10::StrongTypePtr, c10::IValue)> >, c10::optional<c10::Device>, caffe2::serialize::PyTorchStreamReader&) + 637 (0x11aafbe1d in libtorch_cpu.dylib)
frame #7: torch::jit::(anonymous namespace)::ScriptModuleDeserializer::readArchive(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 134 (0x11aafe566 in libtorch_cpu.dylib)
frame #8: torch::jit::(anonymous namespace)::ScriptModuleDeserializer::deserialize(c10::optional<c10::Device>, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > >&) + 1363 (0x11aafca33 in libtorch_cpu.dylib)
frame #9: torch::jit::load(std::__1::shared_ptr<caffe2::serialize::ReadAdapterInterface>, c10::optional<c10::Device>, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > >&) + 477 (0x11aafd95d in libtorch_cpu.dylib)
frame #10: torch::jit::load(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, c10::optional<c10::Device>, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > >&) + 133 (0x11aafdcd5 in libtorch_cpu.dylib)
frame #11: TorchPlugin::TorchForceImpl::initialize(OpenMM::ContextImpl&) + 75 (0x132e7ddeb in libOpenMMTorch.dylib)
frame #12: OpenMM::ContextImpl::initialize() + 1212 (0x10782bdcc in libOpenMM.7.5.dylib)
frame #13: OpenMM::Context::Context(OpenMM::System const&, OpenMM::Integrator&, OpenMM::Platform&) + 121 (0x107828169 in libOpenMM.7.5.dylib)
frame #14: _wrap_new_Context(_object*, _object*) + 678 (0x10756cd06 in _openmm.cpython-39-darwin.so)
<omitting python frames>
frame #43: start + 1 (0x7fff2034c631 in libdyld.dylib)
frame #44: 0x0 + 4 (0x4 in ???)`
Do you know what could be causing the segmentation fault, or the error on my home computer?

**Comment by peastman** [comment]
It looks like that's a known PyTorch bug.  It was reported in https://github.com/pytorch/pytorch/issues/47570, and supposedly was fixed in https://github.com/pytorch/pytorch/pull/48741, which was merged on Dec. 3.  But you're still getting the same error in the latest release?

For the segfault, can you try running it in gdb?  Let it run until it crashes, then use `bt` to get a stack trace for where it crashed.

**Comment by victorprincipe** [comment]
I'm using the following versions of pytorch on my home computer:
```
pytorch                   1.8.0           cpu_py39h4f2e8f6_0    conda-forge
pytorch-cpu               1.8.0           cpu_py39he781eb1_0    conda-forge
```

I've run the code in gdb for the remote workstation that has the segfault, which gave the following output:
```
(gdb) run md_ani_2.py solvated_complex.prmtop solvated_complex.inpcrd 
Starting program: /home/victor/miniconda3/envs/openmmanidev/bin/python md_ani_2.py solvated_complex.prmtop solvated_complex.inpcrd
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
warning: Corrupted shared library list: 0x5555560000b0 != 0x55555600b530

Program received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
```

**Comment by peastman** [comment]
You need to type `bt` to get a stack trace for where the error happened.

**Comment by victorprincipe** [comment]
Ah sorry forgot about that bit, here it is:
```
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x00007fffdb938888 in void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char*>(char*, char*, std::forward_iterator_tag) [clone .constprop.0] () from /home/victor/miniconda3/envs/openmmanidev/lib/plugins/libOpenMMTorchCUDA.so
#2  0x00007fffdb9395aa in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > __gnu_cxx::__to_xstring<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, char>(int (*)(char*, unsigned long, char const*, __va_list_tag*), unsigned long, char const*, ...) () from /home/victor/miniconda3/envs/openmmanidev/lib/plugins/libOpenMMTorchCUDA.so
#3  0x00007fffd61de5d5 in c10::detail::infer_schema::(anonymous namespace)::createArgumentVector(c10::ArrayRef<c10::detail::infer_schema::ArgumentDef>) () from /home/victor/miniconda3/envs/openmmanidev/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so
#4  0x00007fffd61df550 in c10::detail::infer_schema::make_function_schema(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&&, c10::ArrayRef<c10::detail::infer_schema::ArgumentDef>, c10::ArrayRef<c10::detail::infer_schema::ArgumentDef>) ()
   from /home/victor/miniconda3/envs/openmmanidev/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so
#5  0x00007fffd61e1367 in c10::detail::infer_schema::make_function_schema(c10::ArrayRef<c10::detail::infer_schema::ArgumentDef>, c10::ArrayRef<c10::detail::infer_schema::ArgumentDef>) () from /home/victor/miniconda3/envs/openmmanidev/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so
#6  0x00007fffd614270c in std::unique_ptr<c10::FunctionSchema, std::default_delete<c10::FunctionSchema> > c10::detail::inferFunctionSchemaFromFunctor<c10::impl::detail::WrapFunctionIntoRuntimeFunctor_<at::Tensor (*)(at::Tensor const&, at::Tensor const&, c10::optional<at::Tensor> const&, c10::ArrayRef<long>, c10::ArrayRef<long>, c10::ArrayRef<long>, bool, c10::ArrayRef<long>, long, bool, bool, bool), at::Tensor, c10::guts::typelist::typelist<at::Tensor const&, at::Tensor const&, c10::optional<at::Tensor> const&, c10::ArrayRef<long>, c10::ArrayRef<long>, c10::ArrayRef<long>, bool, c10::ArrayRef<long>, long, bool, bool, bool> > >() () from /home/victor/miniconda3/envs/openmmanidev/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so
#7  0x00007fffd614f27f in at::autocast::(anonymous namespace)::TORCH_LIBRARY_IMPL_init_aten_Autocast_2(torch::Library&) ()
   from /home/victor/miniconda3/envs/openmmanidev/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so
#8  0x00007fffd608d300 in torch::detail::TorchLibraryInit::TorchLibraryInit(torch::Library::Kind, void (*)(torch::Library&), char const*, c10::optional<c10::DispatchKey>, char const*, unsigned int) ()
   from /home/victor/miniconda3/envs/openmmanidev/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so
#9  0x00007fffd5f2949f in _GLOBAL__sub_I_autocast_mode.cpp ()
   from /home/victor/miniconda3/envs/openmmanidev/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so
#10 0x00007ffff7de5733 in call_init (env=0x555555a14940, argv=0x7fffffffe238, argc=4, l=<optimised out>) at dl-init.c:72
#11 _dl_init (main_map=main_map@entry=0x55555e2e1e10, argc=4, argv=0x7fffffffe238, env=0x555555a14940) at dl-init.c:119
#12 0x00007ffff7dea1ff in dl_open_worker (a=a@entry=0x7fffffff9ab0) at dl-open.c:522
#13 0x00007ffff792c2df in __GI__dl_catch_exception (exception=0x7fffffff9a90, operate=0x7ffff7de9dc0 <dl_open_worker>, args=0x7fffffff9ab0)
    at dl-error-skeleton.c:196
#14 0x00007ffff7de97ca in _dl_open (
    file=0x7fff2da57200 "/home/victor/miniconda3/envs/openmmanidev/lib/python3.7/site-packages/torch/_C.cpython-37m-x86_64-linux-gnu.so", 
    mode=-2147483646, caller_dlopen=0x55555575eeed <_PyImport_FindSharedFuncptr+141>, nsid=<optimised out>, argc=4, argv=<optimised out>, 
    env=0x555555a14940) at dl-open.c:605
---Type <return> to continue, or q <return> to quit---
#15 0x00007ffff75c1f96 in dlopen_doit (a=a@entry=0x7fffffff9ce0) at dlopen.c:66
#16 0x00007ffff792c2df in __GI__dl_catch_exception (exception=exception@entry=0x7fffffff9c80, operate=0x7ffff75c1f40 <dlopen_doit>, 
    args=0x7fffffff9ce0) at dl-error-skeleton.c:196
#17 0x00007ffff792c36f in __GI__dl_catch_error (objname=0x555555929360, errstring=0x555555929368, mallocedp=0x555555929358, operate=<optimised out>, 
    args=<optimised out>) at dl-error-skeleton.c:215
#18 0x00007ffff75c2735 in _dlerror_run (operate=operate@entry=0x7ffff75c1f40 <dlopen_doit>, args=args@entry=0x7fffffff9ce0) at dlerror.c:162
#19 0x00007ffff75c2051 in __dlopen (file=<optimised out>, mode=<optimised out>) at dlopen.c:87
#20 0x000055555575eeed in _PyImport_FindSharedFuncptr () at /tmp/build/80754af9/python_1614362349910/work/Python/dynload_shlib.c:96
#21 0x0000555555783750 in _PyImport_LoadDynamicModuleWithSpec () at /tmp/build/80754af9/python_1614362349910/work/Python/importdl.c:129
#22 0x00005555557839a9 in _imp_create_dynamic_impl.isra.15 (file=0x0, spec=0x7fff2d3c9110)
    at /tmp/build/80754af9/python_1614362349910/work/Python/import.c:2174
#23 _imp_create_dynamic () at /tmp/build/80754af9/python_1614362349910/work/Python/clinic/import.c.h:289
#24 0x000055555568a6a2 in _PyMethodDef_RawFastCallDict () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:530
#25 0x000055555568a7c1 in _PyCFunction_FastCallDict (func=0x7ffff7f7b7d0, args=<optimised out>, nargs=<optimised out>, kwargs=<optimised out>)
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:585
#26 0x0000555555726bd9 in do_call_core (kwdict=0x7fff33e6c8c0, callargs=0x7fff2dcfbdd0, func=0x7ffff7f7b7d0)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4641
#27 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3191
#28 0x0000555555668d09 in _PyEval_EvalCodeWithName () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3930
#29 0x00005555556b90b7 in _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:433
#30 0x0000555555725ae9 in call_function (kwnames=0x0, oparg=<optimised out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4616
#31 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3093
#32 0x00005555556b8e2b in function_code_fastcall (globals=<optimised out>, nargs=2, args=<optimised out>, co=<optimised out>)
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:283
#33 _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:408
#34 0x00005555557216f0 in call_function (kwnames=0x0, oparg=<optimised out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4616
#35 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3110
#36 0x00005555556b8e2b in function_code_fastcall (globals=<optimised out>, nargs=1, args=<optimised out>, co=<optimised out>)
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:283
#37 _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:408
#38 0x0000555555721466 in call_function (kwnames=0x0, oparg=<optimised out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4616
#39 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3124
---Type <return> to continue, or q <return> to quit---
#40 0x00005555556b8e2b in function_code_fastcall (globals=<optimised out>, nargs=1, args=<optimised out>, co=<optimised out>)
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:283
#41 _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:408
#42 0x0000555555721466 in call_function (kwnames=0x0, oparg=<optimised out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4616
#43 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3124
#44 0x00005555556b8e2b in function_code_fastcall (globals=<optimised out>, nargs=2, args=<optimised out>, co=<optimised out>)
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:283
#45 _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:408
#46 0x0000555555721466 in call_function (kwnames=0x0, oparg=<optimised out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4616
#47 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3124
#48 0x0000555555669d2b in function_code_fastcall (globals=<optimised out>, nargs=2, args=<optimised out>, co=0x7ffff7f70930)
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:283
#49 _PyFunction_FastCallDict () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:322
#50 0x0000555555688a9e in object_vacall () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:1200
#51 0x00005555556e2c0d in _PyObject_CallMethodIdObjArgs () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:1250
#52 0x000055555567248c in import_find_and_load (abs_name=0x7fff2e6a2030) at /tmp/build/80754af9/python_1614362349910/work/Python/import.c:1652
#53 PyImport_ImportModuleLevelObject () at /tmp/build/80754af9/python_1614362349910/work/Python/import.c:1764
#54 0x0000555555723ecf in import_name (level=0x5555558be2e0 <small_ints+160>, fromlist=0x7fff2dd6d5d0, name=0x7fff2e6a2030, f=0x55555b522120)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4770
#55 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:2600
#56 0x0000555555668d09 in _PyEval_EvalCodeWithName () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3930
#57 0x0000555555669be4 in PyEval_EvalCodeEx () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3959
#58 0x0000555555669c0c in PyEval_EvalCode (co=<optimised out>, globals=<optimised out>, locals=<optimised out>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:524
#59 0x0000555555730cb1 in builtin_exec_impl.isra.12 (locals=0x7fff33e6c460, globals=0x7fff33e6c460, source=0x7fff2d5ddf60)
    at /tmp/build/80754af9/python_1614362349910/work/Python/bltinmodule.c:1079
#60 builtin_exec () at /tmp/build/80754af9/python_1614362349910/work/Python/clinic/bltinmodule.c.h:283
#61 0x000055555568a6a2 in _PyMethodDef_RawFastCallDict () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:530
#62 0x000055555568a7c1 in _PyCFunction_FastCallDict (func=0x7ffff7fc9e10, args=<optimised out>, nargs=<optimised out>, kwargs=<optimised out>)
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:585
#63 0x0000555555726bd9 in do_call_core (kwdict=0x7fff33e6c7d0, callargs=0x7fff2d3d5be0, func=0x7ffff7fc9e10)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4641
#64 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3191
---Type <return> to continue, or q <return> to quit---
#65 0x0000555555668d09 in _PyEval_EvalCodeWithName () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3930
#66 0x00005555556b90b7 in _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:433
#67 0x0000555555725ae9 in call_function (kwnames=0x0, oparg=<optimised out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4616
#68 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3093
#69 0x00005555556b8e2b in function_code_fastcall (globals=<optimised out>, nargs=2, args=<optimised out>, co=<optimised out>)
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:283
#70 _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:408
#71 0x00005555557216f0 in call_function (kwnames=0x0, oparg=<optimised out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4616
#72 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3110
#73 0x00005555556b8e2b in function_code_fastcall (globals=<optimised out>, nargs=1, args=<optimised out>, co=<optimised out>)
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:283
#74 _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:408
#75 0x0000555555721466 in call_function (kwnames=0x0, oparg=<optimised out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4616
#76 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3124
#77 0x00005555556b8e2b in function_code_fastcall (globals=<optimised out>, nargs=2, args=<optimised out>, co=<optimised out>)
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:283
#78 _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:408
#79 0x0000555555721466 in call_function (kwnames=0x0, oparg=<optimised out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4616
#80 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3124
#81 0x0000555555669d2b in function_code_fastcall (globals=<optimised out>, nargs=2, args=<optimised out>, co=0x7ffff7f70930)
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:283
#82 _PyFunction_FastCallDict () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:322
#83 0x0000555555688a9e in object_vacall () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:1200
#84 0x00005555556e2c0d in _PyObject_CallMethodIdObjArgs () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:1250
#85 0x000055555567248c in import_find_and_load (abs_name=0x7fff3f7ac330) at /tmp/build/80754af9/python_1614362349910/work/Python/import.c:1652
#86 PyImport_ImportModuleLevelObject () at /tmp/build/80754af9/python_1614362349910/work/Python/import.c:1764
#87 0x0000555555723ecf in import_name (level=0x5555558be2e0 <small_ints+160>, fromlist=0x55555589cf70 <_Py_NoneStruct>, name=0x7fff3f7ac330, 
    f=0x7fff2d4ce050) at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4770
#88 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:2600
#89 0x0000555555668d09 in _PyEval_EvalCodeWithName () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3930
#90 0x0000555555669be4 in PyEval_EvalCodeEx () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3959
---Type <return> to continue, or q <return> to quit---
#91 0x0000555555669c0c in PyEval_EvalCode (co=<optimised out>, globals=<optimised out>, locals=<optimised out>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:524
#92 0x0000555555730cb1 in builtin_exec_impl.isra.12 (locals=0x7fff33e68e60, globals=0x7fff33e68e60, source=0x7fff3f799a50)
    at /tmp/build/80754af9/python_1614362349910/work/Python/bltinmodule.c:1079
#93 builtin_exec () at /tmp/build/80754af9/python_1614362349910/work/Python/clinic/bltinmodule.c.h:283
#94 0x000055555568a6a2 in _PyMethodDef_RawFastCallDict () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:530
#95 0x000055555568a7c1 in _PyCFunction_FastCallDict (func=0x7ffff7fc9e10, args=<optimised out>, nargs=<optimised out>, kwargs=<optimised out>)
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:585
#96 0x0000555555726bd9 in do_call_core (kwdict=0x7fff33e6c1e0, callargs=0x7fff2d3f13c0, func=0x7ffff7fc9e10)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4641
#97 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3191
#98 0x0000555555668d09 in _PyEval_EvalCodeWithName () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3930
#99 0x00005555556b90b7 in _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:433
#100 0x0000555555725ae9 in call_function (kwnames=0x0, oparg=<optimised out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4616
#101 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3093
#102 0x00005555556b8e2b in function_code_fastcall (globals=<optimised out>, nargs=2, args=<optimised out>, co=<optimised out>)
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:283
#103 _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:408
#104 0x00005555557216f0 in call_function (kwnames=0x0, oparg=<optimised out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4616
#105 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3110
#106 0x00005555556b8e2b in function_code_fastcall (globals=<optimised out>, nargs=1, args=<optimised out>, co=<optimised out>)
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:283
#107 _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:408
#108 0x0000555555721466 in call_function (kwnames=0x0, oparg=<optimised out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4616
#109 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3124
#110 0x00005555556b8e2b in function_code_fastcall (globals=<optimised out>, nargs=2, args=<optimised out>, co=<optimised out>)
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:283
#111 _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:408
#112 0x0000555555721466 in call_function (kwnames=0x0, oparg=<optimised out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4616
#113 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3124
#114 0x0000555555669d2b in function_code_fastcall (globals=<optimised out>, nargs=2, args=<optimised out>, co=0x7ffff7f70930)
---Type <return> to continue, or q <return> to quit---
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:283
#115 _PyFunction_FastCallDict () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:322
#116 0x0000555555688a9e in object_vacall () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:1200
#117 0x00005555556e2c0d in _PyObject_CallMethodIdObjArgs () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:1250
#118 0x000055555567248c in import_find_and_load (abs_name=0x7fff31000ef0) at /tmp/build/80754af9/python_1614362349910/work/Python/import.c:1652
#119 PyImport_ImportModuleLevelObject () at /tmp/build/80754af9/python_1614362349910/work/Python/import.c:1764
#120 0x0000555555723ecf in import_name (level=0x5555558be300 <small_ints+192>, fromlist=0x7fff37053fd0, name=0x7ffff4ab9df0, f=0x5555562f1140)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4770
#121 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:2600
#122 0x0000555555668d09 in _PyEval_EvalCodeWithName () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3930
#123 0x0000555555669be4 in PyEval_EvalCodeEx () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3959
#124 0x0000555555669c0c in PyEval_EvalCode (co=<optimised out>, globals=<optimised out>, locals=<optimised out>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:524
#125 0x0000555555730cb1 in builtin_exec_impl.isra.12 (locals=0x7fff33e68960, globals=0x7fff33e68960, source=0x7fff3f788270)
    at /tmp/build/80754af9/python_1614362349910/work/Python/bltinmodule.c:1079
#126 builtin_exec () at /tmp/build/80754af9/python_1614362349910/work/Python/clinic/bltinmodule.c.h:283
#127 0x000055555568a6a2 in _PyMethodDef_RawFastCallDict () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:530
#128 0x000055555568a7c1 in _PyCFunction_FastCallDict (func=0x7ffff7fc9e10, args=<optimised out>, nargs=<optimised out>, kwargs=<optimised out>)
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:585
#129 0x0000555555726bd9 in do_call_core (kwdict=0x7fff33e68be0, callargs=0x7fff33b4b280, func=0x7ffff7fc9e10)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4641
#130 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3191
#131 0x0000555555668d09 in _PyEval_EvalCodeWithName () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3930
#132 0x00005555556b90b7 in _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:433
#133 0x0000555555725ae9 in call_function (kwnames=0x0, oparg=<optimised out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4616
#134 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3093
#135 0x00005555556b8e2b in function_code_fastcall (globals=<optimised out>, nargs=2, args=<optimised out>, co=<optimised out>)
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:283
#136 _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:408
#137 0x00005555557216f0 in call_function (kwnames=0x0, oparg=<optimised out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4616
#138 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3110
#139 0x00005555556b8e2b in function_code_fastcall (globals=<optimised out>, nargs=1, args=<optimised out>, co=<optimised out>)
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:283
---Type <return> to continue, or q <return> to quit---
#140 _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:408
#141 0x0000555555721466 in call_function (kwnames=0x0, oparg=<optimised out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4616
#142 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3124
#143 0x00005555556b8e2b in function_code_fastcall (globals=<optimised out>, nargs=2, args=<optimised out>, co=<optimised out>)
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:283
#144 _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:408
#145 0x0000555555721466 in call_function (kwnames=0x0, oparg=<optimised out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4616
#146 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3124
#147 0x0000555555669d2b in function_code_fastcall (globals=<optimised out>, nargs=2, args=<optimised out>, co=0x7ffff7f70930)
    at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:283
#148 _PyFunction_FastCallDict () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:322
#149 0x0000555555688a9e in object_vacall () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:1200
#150 0x00005555556e2c0d in _PyObject_CallMethodIdObjArgs () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:1250
#151 0x000055555567248c in import_find_and_load (abs_name=0x7fff3f7ac370) at /tmp/build/80754af9/python_1614362349910/work/Python/import.c:1652
#152 PyImport_ImportModuleLevelObject () at /tmp/build/80754af9/python_1614362349910/work/Python/import.c:1764
#153 0x0000555555723ecf in import_name (level=0x5555558be2e0 <small_ints+160>, fromlist=0x55555589cf70 <_Py_NoneStruct>, name=0x7fff3f7ac370, 
    f=0x55555b5207e0) at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4770
#154 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:2600
#155 0x0000555555669640 in _PyEval_EvalCodeWithName () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3930
#156 0x0000555555669df5 in _PyFunction_FastCallDict () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:376
#157 0x00005555556887a3 in _PyObject_Call_Prepend () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:906
#158 0x000055555567b3ae in PyObject_Call () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:245
#159 0x0000555555722e97 in do_call_core (kwdict=0x7fff3a3df870, callargs=0x7fff362db230, func=0x7fff362dc1e0)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4645
#160 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3191
#161 0x0000555555669640 in _PyEval_EvalCodeWithName () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3930
#162 0x00005555556b90b7 in _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1614362349910/work/Objects/call.c:433
#163 0x00005555557216f0 in call_function (kwnames=0x0, oparg=<optimised out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:4616
#164 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3110
#165 0x0000555555668d09 in _PyEval_EvalCodeWithName () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3930
#166 0x0000555555669be4 in PyEval_EvalCodeEx () at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:3959
#167 0x0000555555669c0c in PyEval_EvalCode (co=<optimised out>, globals=<optimised out>, locals=<optimised out>)
---Type <return> to continue, or q <return> to quit---
    at /tmp/build/80754af9/python_1614362349910/work/Python/ceval.c:524
#168 0x0000555555780a74 in run_mod () at /tmp/build/80754af9/python_1614362349910/work/Python/pythonrun.c:1035
#169 0x000055555578ade1 in PyRun_FileExFlags () at /tmp/build/80754af9/python_1614362349910/work/Python/pythonrun.c:988
#170 0x000055555578afd3 in PyRun_SimpleFileExFlags () at /tmp/build/80754af9/python_1614362349910/work/Python/pythonrun.c:429
#171 0x000055555578c105 in pymain_run_file (p_cf=0x7fffffffdfd0, filename=0x5555558c2900 L"md_ani_2.py", fp=0x55555592cf50)
    at /tmp/build/80754af9/python_1614362349910/work/Modules/main.c:462
#172 pymain_run_filename (cf=0x7fffffffdfd0, pymain=0x7fffffffe0e0) at /tmp/build/80754af9/python_1614362349910/work/Modules/main.c:1652
#173 pymain_run_python (pymain=0x7fffffffe0e0) at /tmp/build/80754af9/python_1614362349910/work/Modules/main.c:2913
#174 pymain_main () at /tmp/build/80754af9/python_1614362349910/work/Modules/main.c:3460
#175 0x000055555578c22c in _Py_UnixMain () at /tmp/build/80754af9/python_1614362349910/work/Modules/main.c:3495
#176 0x00007ffff77e6b97 in __libc_start_main (main=0x55555564a0a0 <main>, argc=4, argv=0x7fffffffe238, init=<optimised out>, fini=<optimised out>, 
    rtld_fini=<optimised out>, stack_end=0x7fffffffe228) at ../csu/libc-start.c:310
#177 0x0000555555730e90 in _start () at ../sysdeps/x86_64/elf/start.S:103
```

**Comment by peastman** [comment]
I don't see anything related to OpenMM in that trace.  It's in Python code up to stack frame 20, where Python calls `dlopen()` to load a module.  It stays in `dlopen()` code until stack frame 9, where it enters initialization code for libtorch-cpu.so, and it eventually crashes inside that library.

I would guess this is some sort of version incompatibility, where you somehow have installed different libraries that are either compiled in incompatible ways, or have incompatible versions.

**Comment by victorprincipe** [comment]
Ok thank you for looking into that for me, I will try and compare it to the other environment (that doesn't have the nightly build) and see how versions differ.
Is this also what is causing the separate issue on macOS?

**Comment by peastman** [comment]
>  Is this also what is causing the separate issue on macOS?

I don't think so.  That error is inside the code for loading a TorchScript module, and the issue linked above suggests it's a PyTorch bug.  Make sure you really are using PyTorch 1.8, not any earlier version.

**Comment by victorprincipe** [comment]
Just checked again and it says I'm on version 1.8.0. I'll raise an issue on the PyTorch page and hopefully they can figure out what's going on with that error.

**Comment by peastman** [comment]
It's possible to have multiple versions on your computer, such that conda lists one version, but the OpenMM plugin is actually linking to a different one.  Make sure you don't have an old version of libtorch hanging around somewhere.

**Comment by victorprincipe** [comment]
Turns out that PyTorch v1.8.0 was causing all the errors- I uninstalled miniconda and reinstalled it and all the packages and I had the same issue. After downgrading to Pytorch v1.7.1 I didn't get the error anymore. Now I'm still getting error regarding releasing the GIL when setting velocities to a temperature, even on the nightly build (py39hbb05945_6_apple).

**Comment by victorprincipe** [comment]
Also, on the Linux workstation, I installed pytorch v1.8.0 through pip (as conda didn't allow it due to conflicts), which then gave seg faults so maybe that was what was causing the seg faults previously?

**Comment by peastman** [comment]
You really shouldn't be getting the GIL error if you're using the latest nightly build.  Try running the following script.

```python
from openmm import *
print(version.git_revision)
```

What does it print?

> Turns out that PyTorch v1.8.0 was causing all the errors

Were you using a version of the plugin that was compiled against 1.7.1?  If so, then it's expected not to work with 1.8.  PyTorch doesn't maintain binary compatibility.

**Comment by victorprincipe** [comment]
I had to change the script to:
```
from simtk.openmm import *
print(version.git_revision)
```
And it printed: 
```
b49b82efb5a253a7c891ca084b3370e181de2ea3
```
> Were you using a version of the plugin that was compiled against 1.7.1?
To be honest I'm not sure how to check this... I just noticed that downgrading to v1.7.1 solved the unknown device error.

**Comment by peastman** [comment]
That revision corresponds to 7.5.  You aren't using a nightly build.

**Comment by victorprincipe** [comment]
Hm that's strange, I used 
```conda install -c conda-forge -c omnia-dev openmm```
again just now and it said all packages are up to date. Don't know why it's not installing the latest nightly build...

**Comment by victorprincipe** [comment]
Turns out I could install it on Linux using the `--no-channel-priority` option, however I'm now getting issues during energy minimisation similar to #26 , but I can't install pytorch 1.8.0 through conda and installing it through pip gives segfaults

**Comment by peastman** [comment]
We'll try to get out a PyTorch 1.8 build soon.

**Comment by victorprincipe** [comment]
Hi, I just wanted to let you know that I tried to run the script on the CPU and didn't get segfaults, but I'm still getting the issue with the GIL, even though I've got the nightly build of OpenMM 7.5.1 installed. The script you sent through (below) gave the following output: `98d817304cf97c78d19b7b54304fb0b1ecb90566` 
```
from openmm import *
print(version.git_revision)
```

I'm also using PyTorch 1.7.1

Here is the error that was raised while setting velocities to temperature:

```
Warning: importing 'simtk.openmm' is deprecated.  Import 'openmm' instead.
Minimizing...
Traceback (most recent call last):
  File "md_ani.py", line 40, in <module>
    simulation.context.setVelocitiesToTemperature(300*unit.kelvin)
  File "/home/victor/miniconda3/envs/openmmanidev2/lib/python3.7/site-packages/openmm/openmm.py", line 10515, in setVelocitiesToTemperature
    return _openmm.Context_setVelocitiesToTemperature(self, *args)
openmm.OpenMMException: The autograd engine was called while holding the GIL. If you are using the C++ API, the autograd engine is an expensive operation that does not require the GIL to be held so you should release it with 'pybind11::gil_scoped_release no_gil;'. If you are not using the C++ API, please report a bug to the pytorch team.
Exception raised from execute at /tmp/pip-req-build-h11ydgcy/torch/csrc/autograd/python_engine.cpp:93 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x6a (0x7fac4eae60ea in /home/victor/miniconda3/envs/openmmanidev2/lib/python3.7/site-packages/torch/lib/libc10.so)
frame #1: torch::autograd::python::PythonEngine::execute(std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> > const&, std::vector<at::Tensor, std::allocator<at::Tensor> > const&, bool, bool, std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> > const&) + 0x1cf (0x7fabe638569f in /home/victor/miniconda3/envs/openmmanidev2/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #2: <unknown function> + 0x2b154f0 (0x7fac7b3864f0 in /home/victor/miniconda3/envs/openmmanidev2/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so)
frame #3: torch::autograd::backward(std::vector<at::Tensor, std::allocator<at::Tensor> > const&, std::vector<at::Tensor, std::allocator<at::Tensor> > const&, c10::optional<bool>, bool) + 0x7f (0x7fac7b38722f in /home/victor/miniconda3/envs/openmmanidev2/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so)
frame #4: <unknown function> + 0x30b42f1 (0x7fac7b9252f1 in /home/victor/miniconda3/envs/openmmanidev2/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so)
frame #5: <unknown function> + 0x30b452d (0x7fac7b92552d in /home/victor/miniconda3/envs/openmmanidev2/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so)
frame #6: <unknown function> + 0x1384453 (0x7fac79bf5453 in /home/victor/miniconda3/envs/openmmanidev2/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so)
frame #7: at::Tensor::backward(c10::optional<at::Tensor> const&, c10::optional<bool>, bool) const + 0x100 (0x7fac79bf58d0 in /home/victor/miniconda3/envs/openmmanidev2/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so)
frame #8: TorchPlugin::ReferenceCalcTorchForceKernel::execute(OpenMM::ContextImpl&, bool, bool) + 0xab1 (0x7fac11457b21 in /home/victor/miniconda3/envs/openmmanidev2/lib/plugins/libOpenMMTorchReference.so)
frame #9: OpenMM::ContextImpl::calcForcesAndEnergy(bool, bool, int) + 0xc9 (0x7fac97399d29 in /home/victor/miniconda3/envs/openmmanidev2/lib/python3.7/site-packages/openmm/../../../libOpenMM.so.7.5)
frame #10: OpenMM::Context::setVelocitiesToTemperature(double, int) + 0xd5 (0x7fac973974b5 in /home/victor/miniconda3/envs/openmmanidev2/lib/python3.7/site-packages/openmm/../../../libOpenMM.so.7.5)
frame #11: <unknown function> + 0xff304 (0x7fac97705304 in /home/victor/miniconda3/envs/openmmanidev2/lib/python3.7/site-packages/openmm/_openmm.cpython-37m-x86_64-linux-gnu.so)
<omitting python frames>
```

**Comment by peastman** [comment]
Thanks!  It looks like `setVelocitiesToTemperature()` is another method that needs to release the GIL.  When using a leapfrog integrator, it needs to compute forces to apply the half step offset.

**Comment by victorprincipe** [comment]
I've gotten around the issue by running equilibration using the traditional MM system and then passing the positions/velocities to the new system for the production simulation. Strangely, the production run for about 1800 steps and then this error is raised:
```
Minimizing...
Equilibrating...
Running production...
[W CUDAFunctions.cpp:109] Warning: CUDA initialization: The NVIDIA driver on your system is too old (found version 10020). Please update your GPU driver by downloading and installing a new version from the URL: http://www.nvidia.com/Download/index.aspx Alternatively, go to: https://pytorch.org to install a PyTorch version that has been compiled with your version of the CUDA driver. (function operator())
Traceback (most recent call last):
  File "/home/victor/project2021/ML_bind/productions/test_runs/openmm_ani_test/md_ani_3.py", line 93, in <module>
    production(sys.argv[1], sys.argv[2], positions, velocities)
  File "/home/victor/project2021/ML_bind/productions/test_runs/openmm_ani_test/md_ani_3.py", line 83, in production
    simulation.step(5000)
  File "/home/victor/miniconda3/envs/openmm-torch-pt18/lib/python3.9/site-packages/simtk/openmm/app/simulation.py", line 132, in step
    self._simulate(endStep=self.currentStep+steps)
  File "/home/victor/miniconda3/envs/openmm-torch-pt18/lib/python3.9/site-packages/simtk/openmm/app/simulation.py", line 234, in _simulate
    self._generate_reports(wrapped, True)
  File "/home/victor/miniconda3/envs/openmm-torch-pt18/lib/python3.9/site-packages/simtk/openmm/app/simulation.py", line 256, in _generate_reports
    reporter.report(self, state)
  File "/home/victor/miniconda3/envs/openmm-torch-pt18/lib/python3.9/site-packages/parmed/openmm/reporters.py", line 146, in report
    self._checkForErrors(simulation, state)
  File "/home/victor/miniconda3/envs/openmm-torch-pt18/lib/python3.9/site-packages/parmed/openmm/reporters.py", line 268, in _checkForErrors
    raise ValueError('Energy is NaN') # pragma: no cover
ValueError: Energy is NaN
```

When I used the DCD reporter (instead of the Mdcrd reporter) I got a similar issue at a similar step count but it then said that the particle position is not a number (please see below):
```
Minimizing...
Equilibrating...
Running production...
[W CUDAFunctions.cpp:109] Warning: CUDA initialization: The NVIDIA driver on your system is too old (found version 10020). Please update your GPU driver by downloading and installing a new version from the URL: http://www.nvidia.com/Download/index.aspx Alternatively, go to: https://pytorch.org to install a PyTorch version that has been compiled with your version of the CUDA driver. (function operator())
Traceback (most recent call last):
  File "/home/victor/project2021/ML_bind/productions/test_runs/openmm_ani_test/md_ani_3.py", line 93, in <module>
    production(sys.argv[1], sys.argv[2], positions, velocities)
  File "/home/victor/project2021/ML_bind/productions/test_runs/openmm_ani_test/md_ani_3.py", line 83, in production
    simulation.step(3000)
  File "/home/victor/miniconda3/envs/openmm-torch-pt18/lib/python3.9/site-packages/simtk/openmm/app/simulation.py", line 132, in step
    self._simulate(endStep=self.currentStep+steps)
  File "/home/victor/miniconda3/envs/openmm-torch-pt18/lib/python3.9/site-packages/simtk/openmm/app/simulation.py", line 234, in _simulate
    self._generate_reports(wrapped, True)
  File "/home/victor/miniconda3/envs/openmm-torch-pt18/lib/python3.9/site-packages/simtk/openmm/app/simulation.py", line 256, in _generate_reports
    reporter.report(self, state)
  File "/home/victor/miniconda3/envs/openmm-torch-pt18/lib/python3.9/site-packages/simtk/openmm/app/dcdreporter.py", line 108, in report
    self._dcd.writeModel(state.getPositions(), periodicBoxVectors=state.getPeriodicBoxVectors())
  File "/home/victor/miniconda3/envs/openmm-torch-pt18/lib/python3.9/site-packages/simtk/openmm/app/dcdfile.py", line 125, in writeModel
    raise ValueError('Particle position is NaN')
ValueError: Particle position is NaN
```

Any idea what could be causing these issues?

**Comment by victorprincipe** [comment]
Never mind, I changed the time step to 1 fs and it solved the issue! Seems to be working for now but will have to wait and see...
Thanks again for all the help!

**Comment by victorprincipe** [comment]
Hi Peter,
Now that the script runs without errors, I've been running 6 ns simulations using the ANI potential for the ligand during production simulations. I've been running each simulation on a separate GPU. During equilibration steps, the CPUs are hardly in use and the simulation runs quickly. However, when it reaches the production steps, although it says the process is running on the GPU using the `nvidia-smi` command, I've noticed that processes are running on the CPU too. This results in me only being able to run two simulations at a time, as if I try and run four (there's 4 GPUs in total), they start running very slowly. I was wondering if you have any clue as to why this is and why the CPUs are overloaded, even though I've explicitly stated them to run on the GPUs?

Could this have anything to do with the fact that I'm using the build number 4 which is built against PyTorch 1.8 and CUDA 11, even though the Nvidia CUDA version I'm on is 10.2?

Thanks in advance!

Please see below for outputs of `htop` and `nividia-smi` commands.
<img width="571" alt="Screenshot 2021-04-09 at 16 11 57" src="https://user-images.githubusercontent.com/75172693/114201634-548da580-994e-11eb-8698-5432c4949260.png">
<img width="1245" alt="Screenshot 2021-04-09 at 16 18 55" src="https://user-images.githubusercontent.com/75172693/114202676-5310ad00-994f-11eb-9554-93c0d5d1679d.png">
This is the script I'm using for the production part of the simulation:

```python
def production(prmtop_file, crd_file, positions, velocities):
    prmtop = AmberPrmtopFile(prmtop_file)
    inpcrd = AmberInpcrdFile(crd_file)
    potential = MLPotential('ani2x')
    ml_atoms  = [atom.index for atom in prmtop.topology.atoms() if atom.residue.name == "MOL"]
    mm_system = prmtop.createSystem(nonbondedMethod=PME,
    nonbondedCutoff=1.0*unit.nanometers, constraints=HBonds, rigidWater=True,
    ewaldErrorTolerance=0.0005)
    ml_system = potential.createMixedSystem(prmtop.topology, mm_system, ml_atoms)
    # NPT for production:
    ml_system.addForce(MonteCarloBarostat(1*unit.atmospheres, 300*unit.kelvin, 25))

    integrator = LangevinIntegrator(300*unit.kelvin, 1.0/unit.picoseconds,
    1.0*unit.femtoseconds)
    integrator.setConstraintTolerance(0.00001)
    platform = Platform.getPlatformByName('OpenCL')
    properties = {'OpenCLPrecision': 'mixed', 'OpenCLDeviceIndex': str(deviceindex)}
    simulation = Simulation(prmtop.topology, ml_system, integrator, platform, properties)

    # Set positions from end of equilibration
    simulation.context.setPositions(positions)

    # Set velocities from end of equilibration
    simulation.context.setVelocities(velocities)

    print("Running production...")

    # Simulation reporters
    simulation.reporters.append(DCDReporter('production.dcd', 100))
    simulation.reporters.append(StateDataReporter('production.csv', 100, step = True, potentialEnergy = True,
        kineticEnergy=True, temperature = True, density = True,volume=True, totalEnergy= True, separator='\t'))
    simulation.reporters.append(MdcrdReporter('production.mdcrd', 100))
    simulation.step(6000000)

    # Save final frame to PDB file
    positions = simulation.context.getState(getPositions=True).getPositions()
    PDBFile.writeFile(simulation.topology, positions, open('production.pdb', 'w'))
```

**Comment by peastman** [comment]
You're using OpenCL instead of CUDA.  We only tell PyTorch to use CUDA if you use the OpenMM CUDA platform, so it's probably running on the CPU.

**Comment by victorprincipe** [comment]
Ah ok that makes sense. I've changed the script to use CUDA but I'm getting this error now:
`OpenMMException: Error loading CUDA module: CUDA_ERROR_INVALID_PTX (218)`
Does this mean I need to update the CUDA driver from 10.2 to 11, as I'm using the build that works with PyTorch 1.8 and CUDA 11?

**Comment by peastman** [comment]
Possibly so.  You need a driver that's at least as new as the CUDA toolkit.  But what did you do in the equilibration runs?  If those were running on the GPU, you must have done something differently.

**Comment by victorprincipe** [comment]
For the equilibration runs I wasn't using PyTorch. I was just running energy minimisation and equilibration using AMBER force fields for the whole system due to the GIL issue with the `setVelocitiesToTemperature()` method. After equilibration I passed the positions and velocities to the new hybrid MM/ML system. 
I'll try and update the drivers and hopefully that'll solve the issue. Thanks again!
